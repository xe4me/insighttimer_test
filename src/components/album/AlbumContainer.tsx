import React, {FunctionComponent, memo, useCallback, useEffect, useState} from 'react';
import './Album.scss';
import {fetchPhotos, IPhoto} from './fetchPhotos';
import {fetchAlbum, IAlbum} from './fetchAlbum';
import Album from './Album';


interface IProps {
  albumId:string
}
/* @AlbumContainer
  AlbumContainer is a HOC for Album component.
  AlbumContainer will fetch the data needed for Album and it's children.
* */
const AlbumContainer: FunctionComponent<IProps> = memo(({albumId}) => {

  const [album, setAlbums] = useState<IAlbum | null>(null);
  const [photos, setPhotos] = useState<IPhoto[]>([]);
  useEffect(() => {
    albumId && fetchAlbumAndPhotos(albumId)
  }, [albumId]);


  /* @Method fetchAlbumAndPhotos
   * Will make two http request to fetch the album details and the photos list.
   * @param {string} album_id
* */
  async function fetchAlbumAndPhotos(albumId: string) {
    const albums = await fetchAlbum(albumId);
    setAlbums(albums);
    const photos = await fetchPhotos(albumId);
    setPhotos(photos);
  }


  /* @Method filterPhotos
   * filterPhotos will receive an array of photo_ids and filter the photos based on the matching id
   * @param {string[]} photo_ids
* */
  function filterPhotos(photoIds: string[]) {
    setPhotos((photos) => photos.filter(({id}) => !photoIds.includes(id)))
  }

  const handleHide = useCallback(filterPhotos, []);

  return <Album
      handleHide={handleHide}
      album={album}
      photos={photos}/>;
});
export default AlbumContainer;
