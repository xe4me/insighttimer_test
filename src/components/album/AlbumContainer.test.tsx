import React from "react";
import {mount, ReactWrapper} from "enzyme";
import {act} from "react-dom/test-utils";
import AlbumContainer from './AlbumContainer';
import toJson from 'enzyme-to-json';

jest.mock("./fetchAlbum", () => {
  return {
    __esModule: true,
    fetchAlbum: async () => ({
      "_meta": {
        "success": true,
        "code": 200,
        "message": "OK. Everything worked as expected.",
        "rateLimit": {"limit": 30, "remaining": 29, "reset": 2}
      },
      "result": {
        "id": "3",
        "user_id": "758",
        "title": "In facilis aut aperiam quo qui quia.",
        "_links": {
          "self": {"href": "https://gorest.co.in/public-api/albums/3"},
          "edit": {"href": "https://gorest.co.in/public-api/albums/3"}
        }

      }
    })
  };
});

jest.mock("./fetchPhotos", () => {
  return {
    __esModule: true,
    fetchPhotos: async () => ({
      "_meta": {
        "success": true,
        "code": 200,
        "message": "OK. Everything worked as expected.",
        "totalCount": 2,
        "pageCount": 1,
        "currentPage": 1,
        "perPage": 20,
        "rateLimit": {"limit": 30, "remaining": 28, "reset": 4}
      },
      "result": [{
        "id": "1963",
        "album_id": "3",
        "title": "Suscipit ea quae explicabo. Consequuntur quasi aut similique consequatur ex culpa.",
        "url": "https://lorempixel.com/1024/768/abstract/?17727",
        "thumbnail": "https://lorempixel.com/150/150/abstract/?19973",
        "_links": {
          "self": {"href": "https://gorest.co.in/public-api/photos/1963"},
          "edit": {"href": "https://gorest.co.in/public-api/photos/1963"}
        }
      }, {
        "id": "2569",
        "album_id": "3",
        "title": "Quos ab pariatur molestias eum labore. Commodi excepturi harum nihil nihil id repellendus.",
        "url": "https://lorempixel.com/1024/768/abstract/?10565",
        "thumbnail": "https://lorempixel.com/150/150/abstract/?42724",
        "_links": {
          "self": {"href": "https://gorest.co.in/public-api/photos/2569"},
          "edit": {"href": "https://gorest.co.in/public-api/photos/2569"}
        }
      }]
    })
  };

});

test('should fetch the album and photos and render the children', async function () {
  let wrapper: ReactWrapper;
  await act(async () => {
    wrapper = mount(<AlbumContainer albumId={'2'}/>);
  });
  wrapper.update();
  expect(toJson(wrapper)).toMatchSnapshot()
});
