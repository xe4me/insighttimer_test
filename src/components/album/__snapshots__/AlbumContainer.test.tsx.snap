// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch the album and photos and render the children 1`] = `
<Memo()
  albumId="2"
>
  <Memo()
    album={
      Object {
        "_meta": Object {
          "code": 200,
          "message": "OK. Everything worked as expected.",
          "rateLimit": Object {
            "limit": 30,
            "remaining": 29,
            "reset": 2,
          },
          "success": true,
        },
        "result": Object {
          "_links": Object {
            "edit": Object {
              "href": "https://gorest.co.in/public-api/albums/3",
            },
            "self": Object {
              "href": "https://gorest.co.in/public-api/albums/3",
            },
          },
          "id": "3",
          "title": "In facilis aut aperiam quo qui quia.",
          "user_id": "758",
        },
      }
    }
    handleHide={[Function]}
    photos={
      Object {
        "_meta": Object {
          "code": 200,
          "currentPage": 1,
          "message": "OK. Everything worked as expected.",
          "pageCount": 1,
          "perPage": 20,
          "rateLimit": Object {
            "limit": 30,
            "remaining": 28,
            "reset": 4,
          },
          "success": true,
          "totalCount": 2,
        },
        "result": Array [
          Object {
            "_links": Object {
              "edit": Object {
                "href": "https://gorest.co.in/public-api/photos/1963",
              },
              "self": Object {
                "href": "https://gorest.co.in/public-api/photos/1963",
              },
            },
            "album_id": "3",
            "id": "1963",
            "thumbnail": "https://lorempixel.com/150/150/abstract/?19973",
            "title": "Suscipit ea quae explicabo. Consequuntur quasi aut similique consequatur ex culpa.",
            "url": "https://lorempixel.com/1024/768/abstract/?17727",
          },
          Object {
            "_links": Object {
              "edit": Object {
                "href": "https://gorest.co.in/public-api/photos/2569",
              },
              "self": Object {
                "href": "https://gorest.co.in/public-api/photos/2569",
              },
            },
            "album_id": "3",
            "id": "2569",
            "thumbnail": "https://lorempixel.com/150/150/abstract/?42724",
            "title": "Quos ab pariatur molestias eum labore. Commodi excepturi harum nihil nihil id repellendus.",
            "url": "https://lorempixel.com/1024/768/abstract/?10565",
          },
        ],
      }
    }
  >
    <form
      className="Album"
      onSubmit={[Function]}
    >
      <div
        className="Album__Header"
      >
        <Typography
          variant="h5"
        >
          <h5
            className="Typography Typography--h5"
          />
        </Typography>
        <Memo()
          color="secondary"
          type="submit"
        >
          <button
            className="Button Button--secondary"
            type="submit"
          >
            Hide
          </button>
        </Memo()>
      </div>
      <Memo()
        photos={
          Object {
            "_meta": Object {
              "code": 200,
              "currentPage": 1,
              "message": "OK. Everything worked as expected.",
              "pageCount": 1,
              "perPage": 20,
              "rateLimit": Object {
                "limit": 30,
                "remaining": 28,
                "reset": 4,
              },
              "success": true,
              "totalCount": 2,
            },
            "result": Array [
              Object {
                "_links": Object {
                  "edit": Object {
                    "href": "https://gorest.co.in/public-api/photos/1963",
                  },
                  "self": Object {
                    "href": "https://gorest.co.in/public-api/photos/1963",
                  },
                },
                "album_id": "3",
                "id": "1963",
                "thumbnail": "https://lorempixel.com/150/150/abstract/?19973",
                "title": "Suscipit ea quae explicabo. Consequuntur quasi aut similique consequatur ex culpa.",
                "url": "https://lorempixel.com/1024/768/abstract/?17727",
              },
              Object {
                "_links": Object {
                  "edit": Object {
                    "href": "https://gorest.co.in/public-api/photos/2569",
                  },
                  "self": Object {
                    "href": "https://gorest.co.in/public-api/photos/2569",
                  },
                },
                "album_id": "3",
                "id": "2569",
                "thumbnail": "https://lorempixel.com/150/150/abstract/?42724",
                "title": "Quos ab pariatur molestias eum labore. Commodi excepturi harum nihil nihil id repellendus.",
                "url": "https://lorempixel.com/1024/768/abstract/?10565",
              },
            ],
          }
        }
      >
        <ul
          className="Photos"
        />
      </Memo()>
    </form>
  </Memo()>
</Memo()>
`;
